### PR TITLE
Dynamic Warp Size to enable Navi 2x/3x support

### PR DESCRIPTION
**Pull request by xinyazhang**
Fixes https://github.com/ROCmSoftwarePlatform/frameworks-internal/issues/4521

### CODE REVIEW

**Comment by xinyazhang** [comment]
This commit passed the regression test on `gfx90a` with the following command
```
pytest python/test/unit/language/test_core_amd.py --verbose
```

Result:
```
4532 passed, 58 skipped, 143 warnings in 1215.18s (0:20:15)
```

File: lib/Dialect/TritonGPU/IR/Dialect.cpp
Comment (zhanglx13) [inline lib/Dialect/TritonGPU/IR/Dialect.cpp:104]: I think it's better to get the warpSize from the mod, since the layout is usually figured from num_warps and warpSize.

File: include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td
Comment (zhanglx13) [inline include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td:37]: Does this mean the default value is set to 32?
I'd suggest to set 64 as the default value, since MI series are more important targets.

Comment (xinyazhang) [reply include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td:37]: I'm keeping it untouched intentionally, otherwise we will have one extra conflict to deal with whenever sync with upstreams.

Additionally I don't think keeping the default be 32 to accommodate MI is a better approach because
1. The upstream is unlikely to accept 64 as default
2. The divergence makes the behavior of ROCM/triton different with upstream
3. New AMD GPU users need to take extra care NOT using the upstream otherwise their MI GPU system would potentially fail on Triton due to the divergence.

To avoid this we need to make sure Triton works on MI with 32 as default at official branch, rather than forcing potential new users seeking for solutions from some fork. Setting default to 64 only hides potential problems away from users.

File: lib/Dialect/TritonGPU/IR/Dialect.cpp
Comment (xinyazhang) [reply lib/Dialect/TritonGPU/IR/Dialect.cpp:104]: True, but you don't always have mod available.
The ModuleOp is only visible in a few files and the majority part of Triton backend need other sources to figure out the current warp size. `Attribute` is the existing way to propagate the property.

Comment (zhanglx13) [reply lib/Dialect/TritonGPU/IR/Dialect.cpp:104]: Yeah. But you can always obtain the mod easily, such as getOperation() and getParentOfType<ModuleOp>().
Why they need other sources to figure out the warpSize? I think it should be determined as early as possible since it is basically arch-specific.

File: include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td
Comment (zhanglx13) [reply include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td:37]: I dont think MI gpus work with warpSize = 32 when it comes mfma instructions. 
yes, in terms of compatibility, we dont want to diverge from upstream. And some day we also want to live with nvidia gpus in upstream. 
But for now, most of critical applications are running on MI gpu. And a default value of 32 means that we need to change for those applications. 
@jayfurmanek what do you think?

Comment (xinyazhang) [reply include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td:37]: The whole point of the PR is to make sure the warp size can change without re-compiling Triton package, and theoretically on a system with both Navi 31 and MI200 installed, we can run Triton on both GPUs at the same time without installing two copies of Triton.

If MI failed with warp size defaults to 32, it means the Triton JIT compiler has problems figuring out the right warp size for the target architecture.

Comment (jayfurmanek) [reply include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td:37]: >>To avoid this we need to make sure Triton works on MI with 32 as default at official branch
@xinyazhang Do you mean let 32 be the default but do a dynamic check/set based on GPU type?

It needs to be 64 for the MI GPUs at the end of the day.

Comment (xinyazhang) [reply include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td:37]: Another solution is to make the default warp size be a compile time macro and configurable at compile time.

Comment (xinyazhang) [reply include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td:37]: > Do you mean let 32 be the default but do a dynamic check/set based on GPU type?

Yes, this is the whole point of this PR.
The detection logic is in `./python/triton/compiler/compiler.py`. Ultimately it calls `hipGetDeviceProperties` and passes the warpSize to the constructor of the Triton JIT compiler.

Comment (jayfurmanek) [reply include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td:37]: ok - I'll look through it closer. Is there even a scenario where the default is used? Maybe only if the arch is unrecognized?

Comment (xinyazhang) [reply include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td:37]: > ok - I'll look through it closer. Is there even a scenario where the default is used? Maybe only if the arch is unrecognized?

A tl;dr version is the default value should never be used at runtime., but the story about warp size is a little long. 

1. We want the warpSize to be a dynamical property, and hence it should be a property specific to some MLIR pass(es)    
2. This demands leads to the **UPSTREAM** code change at [`include/triton/Conversion/TritonToTritonGPU/Passes.td`](https://github.com/ROCmSoftwarePlatform/triton/pull/244/files#diff-e219b4a9f8bdfbd57b2134863508395eeed113eb9fec36811d055aed9938f68b), which create a new property called threadsPerWarp
3. However there is no direct way to change the property programmatically. The only documented way is using command line options but we don't use it in that way
4. In practice, `class ConvertTritonToTritonGPU` (at file `lib/Conversion/TritonToTritonGPU/TritonToTritonGPUPass.cpp`) inherits `ConvertTritonToTritonGPUBase<ConvertTritonToTritonGPU>` and properly configures the warpSize in its constructor. This constructor is invoked by `createConvertTritonToTritonGPUPass` which is exported as `add_convert_triton_to_tritongpu_pass` in python. This python interface is called by `compiler.py` according to the warpSize queried from `hipGetDeviceProperties`
5. However, there is another function `createConvertTritonToTritonGPUPass()` which doesn't take any argument and creates an `ConvertTritonToTritonGPU` object with default settings, leading to errors. This function should never be called at runtime, but I cannot remove it because this function is required in `registerConvertTritonToTritonGPU()` (generated as `./python/build/cmake.linux-x86_64-cpython-3.10/include/triton/Conversion/TritonToTritonGPU/Passes.h.inc`). My knowledge about MLIR is not enough to remove this entrance potentially in danger.
6. Other properties are fundamentally propagated from `ConvertTritonToTritonGPU` object. An example propagation path is `threadsPerWarp = triton::gpu::TritonGPUDialect::getThreadsPerWarp(mod)->decomposeMmaToDotOperand(..., decomposeMmaToDotOperand)->layout = triton::gpu::BlockedEncodingAttr::get(..., threadsPerWarp)->...->getWarpSize(layout)`

Comment (zhanglx13) [reply include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td:37]: Thanks @xinyazhang for such a nice clarification about warpSize.
Now I see where it comes from:
- When the jit compiler is invoked, i.e. `python ./some_test.py`, warpSize is determined by arch, which is determined by hipGetDeviceProperties. In this case, we do not need to set warpSize.
- When triton-opt is used to lower IRs, warpSize can be set as the parameter. E.g. `triton-opt –convert-triton-to-tritongpu=threadsPerWarp=64`. If this parameter is not set, the default value is 32 in this PR and also upstream

Both paths will go through `ConvertTritonToTritonGPU` and set `triton_gpu.threads-per-warp` as an attribute of the module. Therefore, later passes can call `triton::gpu::TritonGPUDialect::getThreadsPerWarp(mod)` to obtain warpSize.
I’d argue that we change this default value to 64 in favor of MI gpus (at least for now) so that when we use triton-opt on MI gpus,
- we do not need to set the warpSize explicitly as the parameter
- we do not need to include triton_gpu.threads-per-warp in the module. Therefore, we do not need to change existing IRs

Comment (xinyazhang) [reply include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td:37]: > * we do not need to set the warpSize explicitly as the parameter
> * we do not need to include triton_gpu.threads-per-warp in the module. Therefore, we do not need to change existing IRs

Once again this PR is for Navi 2x/3x support, and both changes are needed unless there are better propagation approaches.
If MI is considered as the only target that need to take care of I can then close the PR.

Comment (zhanglx13) [reply include/triton/Dialect/TritonGPU/IR/TritonGPUDialect.td:37]: I think this PR does a great job in enabling Navi gpus. And I also agree that detecting arch at runtime and set warpSize accordingly is an elegant solution to the problem. From this angle, the default value of `threadsPerWarp` does not make any difference.

However, what I am trying to bring up is the scenario where `triton-opt` is used. And the default value does make a difference here.

> Once again this PR is for Navi 2x/3x support, and both changes are needed unless there are better propagation approaches.

Both changes are needed only on MI gpus.

> If MI is considered as the only target that needs to take care of I can then close the PR.

Changing the default to 64 does not mean MI is the only target. Just as keeping it as 32 does not mean Navi is the only target.

> Update: oh I forgot about one more thing. All changes in https://github.com/ROCmSoftwarePlatform/triton/commit/b95482d9ff298b3c5bfaba3677331499dd1fb9c6 come from the Triton upstream directly. I'm not the right person to argue with.

NVIDIA gpus are the main hw for upstream, hence 32 is a valid choice as the default value. So nobody is going to argue that with upstream people.

**Comment by xinyazhang** [comment]
The PR has been updated to match the old behavior of `triton-opt`. @zhanglx13 Can you take a look at if `triton-opt` is affected with this PR? I'm running GPU UTs but they cannot discover behavior changes in the command line tools.

The code resides at branch `xinyazhang/navi-warpsize-32`.

**Comment by zhanglx13** [comment]
We still want the set the default value as `TRITON_GPU_DEFAULT_WARPSIZE` in TritonGPUDialect.td. The old IRs does not have `triton_gpu.threads-per-warp` set in the attributes of the module. And a default value of 32 will lead to incorrect computations when triton-opt is invoked.

Another thing. Why cant we use the `USE_ROCM` guard in td files like here https://github.com/ROCmSoftwarePlatform/triton/blob/ac1e551e8a51ad5969325bdb5a0e75010dff70aa/include/triton/Dialect/TritonGPU/IR/TritonGPUAttrDefs.td#L82

**Comment by xinyazhang** [comment]
PR updated.

> Another thing. Why cant we use the `USE_ROCM` guard in td files like here

Yes we can. I'm just avoiding magic numbers.

File: lib/Dialect/TritonGPU/IR/Dialect.cpp
Comment (zhanglx13) [reply lib/Dialect/TritonGPU/IR/Dialect.cpp:104]: @xinyazhang The upstream does not obtain warpSize from layouts. And we should be able to copy what upstream does. Are there any scenarios that we cannot obtain the mod and then the warpSize from the mod?

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h
Comment (xinyazhang) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h:723]: Sure

File: lib/Dialect/TritonGPU/IR/Dialect.cpp
Comment (xinyazhang) [reply lib/Dialect/TritonGPU/IR/Dialect.cpp:104]: The other method used by upstream is to propagate the warp size through `TritonGPUTypeConverter`, but it's incomplete. Upstream PR https://github.com/openai/triton/pull/1753 is needed to make the remaining warpSize dynamic, but that PR is unmerged yet.

Additionally layout is generally available and the unmerged upstream PR requires more changes on function signatures.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h
Comment (xinyazhang) [reply lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h:718]: Pushed, the latest commit is supposed to get squashed with "Add configurable wavefront size support for Navi/MI." before merge the PR.

File: lib/Dialect/TritonGPU/IR/Dialect.cpp
Comment (zhanglx13) [reply lib/Dialect/TritonGPU/IR/Dialect.cpp:104]: I'm fine to merge this PR for now with an agreement to get rid of this getWarpSize-from-layout scheme when upstream's work about dynamic warpSize, including renaming and better propagation, is done.

**Review by zhanglx13 — APPROVED** [review]

**Comment by xinyazhang** [comment]
`pytest python/test/unit/language/test_core_amd.py --verbose` testing results:

On MI 200: `4532 passed, 58 skipped, 143 warnings in 1223.80s (0:20:23)`
On Navi 31: All tests passed before abortion at 98% due to `python/test/unit/language/test_core_amd.py::test_convert2d[dst_layout0-src_layout1-float16-shape0] Fatal Python error: Aborted`.

**Comment by scxiao** [comment]
> `pytest python/test/unit/language/test_core_amd.py --verbose` testing results:
> 
> On MI 200: `4532 passed, 58 skipped, 143 warnings in 1223.80s (0:20:23)` On Navi 31: All tests passed before abortion at 98% due to `python/test/unit/language/test_core_amd.py::test_convert2d[dst_layout0-src_layout1-float16-shape0] Fatal Python error: Aborted`.

The failed test case here is also due to that the warp size is 32. 
If you look at the layout input for the test_convert2d, they have warpsize 64 (second list in each BlockedLayout). The external guy just used the layout input in the else branch (supposed to be used by cuda test cases, which has warpsize 32), so all tests passed in his code changes.

**Comment by xinyazhang** [comment]
> I think the failed test case here is also due to that the warp size is 32.

You are right. I added another commit to address the reduce2d failures.
The abortion is due to warp size mismatching.
I also changed other layouts because some existing combinations would run out of shared memory.

If you're fine about the changes of UT data I'll merge the PR afterwards.

**Comment by scxiao** [comment]
Just wondering, do you have an idea of why the offline test still fail? Seems like it is also related to the warp_size 32. 
Also, I am wondering whether github let you merge this PR if the offline test failed...

File: test/Conversion/triton_to_tritongpu.mlir
Comment (scxiao) [inline test/Conversion/triton_to_tritongpu.mlir:4]: I ran this test locally, to make this test pass, we need "triton_gpu.threads-per-warp" = 64

File: test/Conversion/triton_to_tritongpu.mlir
Comment (scxiao) [inline test/Conversion/triton_to_tritongpu.mlir:38]: Same as above, to make this test pass, lines 36 to line 38 needs no changes, and in line 39, "triton_gpu.threads-per-warp" = 64.

File: test/Conversion/triton_to_tritongpu.mlir
Comment (xinyazhang) [reply test/Conversion/triton_to_tritongpu.mlir:4]: They should match the architecture's warp size. However I did not run tests related to this file apparently on MI200 so I didn't touch it.

File: test/Conversion/triton_to_tritongpu.mlir
Comment (xinyazhang) [reply test/Conversion/triton_to_tritongpu.mlir:38]: A temporary solution is to have a private patch for these files on our repo.
A proper solution requires these file support dynamic warp size as well, but this is non-trivial.

Maybe we can address this in a separate PR.

File: test/Conversion/triton_to_tritongpu.mlir
Comment (scxiao) [reply test/Conversion/triton_to_tritongpu.mlir:4]: I cannot find it either, so I just ran `root@ubb4-rack-07:/workspace/projects/triton/python# ./build/cmake.linux-x86_64-cpython-3.8/bin/triton-opt  ../test/Conversion/triton_to_tritongpu.mlir -split-input-file -convert-triton-to-tritongpu=num-warps=2`, the did a manual check

File: test/Conversion/triton_to_tritongpu.mlir
Comment (scxiao) [reply test/Conversion/triton_to_tritongpu.mlir:38]: I think my point is make what ever changes (or no changes) to make this PR pass the CI build, so we can get it in. Then on navi31, it is acceptable that this individual test failed (we know the reason of the failure, then we make another PR later to address). 

Seems like github does not let the PR merge if CI build failed, so we need to make this test pass. 
If you do not mind, I can change this test so it can pass the CI build.

Comment (xinyazhang) [reply test/Conversion/triton_to_tritongpu.mlir]: > Seems like github does not let the PR merge if CI build failed

Nope I can get it merged after one approval. I'm just being extra cautious.

**Review by scxiao — APPROVED** [review]

**Comment by scxiao** [comment]
Now the offline tests passed, hopefully we can merge it after CI build is done. Just approved.

**Comment by xinyazhang** [comment]
The workflow runs out of space and I deleted some of them.

The new offline test can be found at: https://github.com/ROCmSoftwarePlatform/triton/actions/runs/5406142295

**Comment by scxiao** [comment]
Hi @xinyazhang, you can merge this PR in now.

**Comment by xinyazhang** [comment]
@scxiao @zhanglx13 merged. I intentionally used "rebase and merge" after squashing tiny changes locally.

Commit 36f7c9872a72e4ed9bc1a73a6a46b0188363577f is kept standalone ([ROCM private]) so it can be removed when submitting PRs to the upstream.

**Comment by BruceDai003** [comment]
https://github.com/ROCmSoftwarePlatform/triton/commit/36f7c9872a72e4ed9bc1a73a6a46b0188363577f#r120480057
Seems this should be `CMAKE_CXX_FLAGS`, otherwise won't build on NV as I just tried.

**Pull request merged**
